src/cargo/sources/registry/mod.rs

```diff
@@ -445,26 +548,17 @@ impl<'cfg> Source for RegistrySource<'cfg> {
     }
 }
 
-// TODO: this is pretty unfortunate, ideally we'd use `DeserializeSeed` which
-//       is intended for "deserializing with context" but that means we couldn't
-//       use `#[derive(Deserialize)]` on `RegistryPackage` unfortunately.
-//
-// I'm told, however, that https://github.com/serde-rs/serde/pull/909 will solve
-// all our problems here. Until that lands this thread local is just a
-// workaround in the meantime.
-//
-// If you're reading this and find this thread local funny, check to see if that
-// PR is merged. If it is then let's ditch this thread local!
-thread_local!(static DEFAULT_ID: RefCell<Option<SourceId>> = Default::default());
-
-impl<'de> de::Deserialize<'de> for DependencyList {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+struct DependencyListDefault(SourceId);
+
+impl<'de> de::DeserializeSeed<'de> for DependencyListDefault {
+    type Value = DependencyList;
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
     where
         D: de::Deserializer<'de>,
     {
-        return deserializer.deserialize_seq(Visitor);
+        return deserializer.deserialize_seq(Visitor(self.0));
```

Comment:
Ditto here, I would use the same struct as both the DeserializeSeed and the Visitor.